scalyr_agent/builtin_monitors/syslog_monitor.py

```diff
@@ -570,77 +583,110 @@ class SyslogUDPHandler(six.moves.socketserver.BaseRequestHandler):
+
+    def factory_method(request_processing_executor):
+        def create_handler(request, client_address, server):
+            return SyslogUDPHandler(request_processing_executor, request, client_address, server)
+
+        return create_handler
+
+
     def handle(self):
         data = six.ensure_text(self.request[0].strip(), "utf-8", errors="ignore")
         extra = {
             "proto": "udp",
             "srcip": self.client_address[0],
             "destport": self.server.server_address[1],
         }
-        self.server.syslog_handler.handle(data, extra)
 
+        self.__request_processing_executor.submit(
+            self.server.syslog_handler.handle, data, extra
+        )
 
-class SyslogRequestParser(object):
-    """
-    Syslog TCP request data parser which supports framed and line delimited syslog messages.
 
-    It output line / frame data to disk (aka calls scalyr logger method which does that) as soon
-    as it's processed.
+class SocketNotReadyException(Exception):
+    def __init__(self, cause):
+        self.cause = cause
 
-    This approach of calling scalyr logger for each frame / line is very inefficient and only allows
-    syslog monitors to achieve a throughput of 1.5-3 MB/s or so.
-    """
+    def __str__(self):
+        return "SocketNotReadyException: " + str(self.cause)
 
-    def __init__(
-        self,
-        socket,
-        socket_client_address,
-        socket_server_address,
-        max_buffer_size,
-        message_size_can_exceed_tcp_buffer=False,
-    ):
-        self._socket = socket
-        self._client_address = socket_client_address
-        self._server_address = socket_server_address
-
-        if socket:
-            self._socket.setblocking(False)
+class SocketClosed(Exception):
+    pass
 
-        self._remaining = None
+class SyslogRequest(object):
+    def __init__(self, socket, max_buffer_size):
+        self._socket = socket
         self._max_buffer_size = max_buffer_size
-        self._message_size_can_exceed_tcp_buffer = message_size_can_exceed_tcp_buffer
-
+        self._socket.setblocking(True)
         self.is_closed = False
 
     def read(self):
         """Reads self._max_buffer_size bytes from the buffer"""
 
-        data = None
         try:
             data = self._socket.recv(self._max_buffer_size)
             if not data:
                 self.is_closed = True
-        except socket.timeout:
-            self._socket_error = True
-            return None
-        except NON_BLOCKING_SOCKET_DATA_NOT_READY_EXCEPTIONS:
-            return None
+                raise SocketClosed()
+        except (socket.timeout, *NON_BLOCKING_SOCKET_DATA_NOT_READY_EXCEPTIONS) as e:
+            raise SocketNotReadyException(e)
         except socket.error as e:
             if e.errno == errno.EAGAIN:
-                return None
+                raise SocketNotReadyException(e)
```

Comment:
Just curious why this change is needed? In the past, we had bug related to this code path throwing, but this could have changed with the multi threading model.

Reply:
Before the method returned None for both cases - when data on socket were not ready and when the socket was closed.
And you had to guess what happened. 
The exceptions are path of the code flow and are expected and catched right away, avoiding any propagating.

Reply:
And I don't like returning None when something unexpected happened. It's quite error-prone.